gustaf/io/hmascii.py

```diff
@@ -0,0 +1,315 @@
+                    )
+
+                # read component
+                elif line.name == "component":
+                    current_component = HMComponent(line)
+                    self.components.append(current_component)
+
+                # read element
+                elif line.name in HMComponent.element_types:
+                    if not current_component:
+                        raise RuntimeError(
+                            "Encountered element before first " "component."
+                        )
+                    current_component.add_element(line)
+
+
+def load(fname, element_type=""):
+    """hmascii load.
+
+    Parameters
+    -----------
+    fname: str
+        The filename ending in '.hmascii'.
+    element_type: str
+        The volume element type to be imported. If this is not specified, the
+        most common element type in the main component is chosen.
+
+    Returns
+    -----------
+    mesh: Mesh
+    """
+    f_name = pathlib.Path(fname)
+
+    if not f_name.suffix.endswith(".hmascii"):
+        raise TypeError("Input file must be of Type .hmascii .")
+
+    hm_model = HMModel(fname)
+
+    if not element_type:
+        # which element types occur in the mesh?
+        element_types_in_model = [
+            element_type
+            for component in hm_model.components
+            for element_type in component.elements
+        ]
+        preferred_element_types = [
+            element_type
+            for element_type in HMComponent.element_type_preference
+            if element_type in element_types_in_model
+        ]
+        if len(preferred_element_types) < 1:
+            raise RuntimeError(
+                "Couldn't find any usable element types in " "model."
+            )
+
+        element_type = preferred_element_types[0]
+        utils.log.info(f"Selected volume element type '{element_type}'.")
+
+    # determine subelement type
+    subelement_type = HMComponent.element_types[element_type].subelement
+
+    hm_volume_elements_nonunique = np.ndarray(
+        shape=(0, HMComponent.element_types[element_type].number_of_nodes),
+        dtype=int,
+    )
+
+    bcs = dict()
+
+    # loop over all components to find volume elements
+    for hm_component in hm_model.components:
+        # can we use all elements?
+        ignored_element_types = set(hm_component.elements).difference(
+            {element_type, subelement_type}
+        )
+        if len(ignored_element_types) > 0:
+            utils.log.warning(
+                f"Component '{hm_component.name}' contains "
+                f"unknown element types {ignored_element_types}. "
+                "They will be ignored."
+            )
+
+        # are there volume elements? append.
+        if element_type in hm_component.elements:
+            elements_in_component = hm_component.elements[element_type]
+
+            # append elements
+            hm_volume_elements_nonunique = np.concatenate(
```

Comment:
this is quite expensive to have in a loop. can you just append to a list and concatenate once at the end?